sdks/java/io/hadoop-input-format/src/main/java/org/apache/beam/sdk/io/hadoop/inputformat/HadoopInputFormatIO.java

```diff
@@ -0,0 +1,1121 @@
+     * pair returned by RecordReader. User provided key and value classes are validated against
+     * parameterized type of InputFormat. If parameterized type has any type parameter such as T,
+     * K, V, etc then validation is done by encoding and decoding first pair returned by
+     * RecordReader.
+     */
+    private <T> boolean validateClasses(
+        String inputFormatGenericType,
+        Coder<T> coder,
+        String property) throws IOException, InterruptedException {
+      Class<?> inputClass = null;
+      try {
+        inputClass = Class.forName(inputFormatGenericType);
+      } catch (Exception e) {
+        /*
+         * Given inputFormatGenericType is a type parameter i.e. T, K, V, etc. In such cases class
+         * validation for user provided input key/value will not work correctly. Therefore the need
+         * to validate key/value classes by encoding and decoding key/value object with the given
+         * coder.
+         */
+        return validateClassUsingCoder(property, coder);
+      }
+      /*
+       * Validates key/value class with InputFormat's parameterized type.
+       */
+      if (inputClass != null) {
+        return validateClassesEquality(inputClass, property);
+      }
+      return true;
+    }
+
+    /**
+     * Returns true if first record's key/value encodes and decodes successfully. Also sets
+     * expectedKeyClass/expectedValueClass if cloning of key/value fails.
+     */
+    private <T> boolean validateClassUsingCoder(String property, Coder<T> coder)
+        throws IOException, InterruptedException {
+      final RecordReader<?, ?> reader;
+      reader = fetchFirstRecord();
+      if (property.contains("key")) {
+        boolean isKeyClassValide = checkEncodingAndDecoding(coder, (T) reader.getCurrentKey());
+        if (!isKeyClassValide) {
+          this.setExpectedKeyClass(reader.getCurrentKey().getClass());
+        }
+        return isKeyClassValide;
+      } else {
+        boolean isValueClassValide = checkEncodingAndDecoding(coder, (T) reader.getCurrentValue());
+        if (!isValueClassValide) {
+          this.setExpectedValueClass(reader.getCurrentValue().getClass());
+        }
+        return isValueClassValide;
+      }
+    }
+
+    /**
+     * Returns true if the class set in a given property by the pipeline user is type of given
+     * expectedClass.
+     */
+    private boolean validateClassesEquality(Class<?> expectedClass, String property) {
+      Class<?> actualClass = conf.getHadoopConfiguration().getClass(property, Object.class);
+      if (!actualClass.isAssignableFrom(expectedClass)) {
+        if (property.contains("key")) {
+          this.setExpectedKeyClass(expectedClass);
+        } else {
+          this.setExpectedValueClass(expectedClass);
+        }
+        return false;
+      }
+      return true;
+    }
+
+    /**
+     * Returns false if exception occurs while encoding and decoding the given value with the given
```

Comment:
Updated the Javadoc comments. Also, have caught CoderException instead of catching the generic Exception.